File: epi_judge_cpp/lowest_common_ancestor_close_ancestor.cc
#include <memory>

#include "binary_tree_with_parent_prototype.h"
#include "test_framework/binary_tree_utils.h"
#include "test_framework/test_failure_exception.h"
#include "test_framework/test_timer.h"

using std::unique_ptr;

BinaryTreeNode<int>* LCA(const unique_ptr<BinaryTreeNode<int>>& node0,
                         const unique_ptr<BinaryTreeNode<int>>& node1) {
  // Implement this placeholder.
  return nullptr;
}

int LcaWrapper(TestTimer& timer, const unique_ptr<BinaryTreeNode<int>>& tree,
               int node0, int node1) {
  timer.Start();
  auto result = LCA(MustFindNode(tree, node0), MustFindNode(tree, node1));
  timer.Stop();

  if (!result) {
    throw TestFailureException("Result can not be nullptr");
  }
  return result->data;
}

#include "test_framework/test_utils_generic_main.h"

int main(int argc, char* argv[]) {
  std::vector<std::string> param_names{"timer", "tree", "node0", "node1"};
  generic_test_main(argc, argv, param_names, "lowest_common_ancestor.tsv",
                    &LcaWrapper);
  return 0;
}

File: epi_judge_python/lowest_common_ancestor_close_ancestor.py
from test_framework.binary_tree_utils import must_find_node
from test_framework.test_failure_exception import TestFailureException
from test_framework.test_utils import enable_timer_hook


def lca(node0, node1):
    # Implement this placeholder.
    return None


@enable_timer_hook
def lca_wrapper(timer, tree, node0, node1):
    timer.start()
    result = lca(must_find_node(tree, node0), must_find_node(tree, node1))
    timer.stop()

    if result is None:
        raise TestFailureException("Result can't be None")
    return result.data


from test_framework import test_utils_generic_main, test_utils

if __name__ == '__main__':
    test_utils_generic_main.generic_test_main('lowest_common_ancestor.tsv',
                                              lca_wrapper)
